### PR TITLE
Fix missing parentheses in inline variable

### LINKED ISSUE #54736: 'Inline variable' should add parens when needed
# Bug Report

### 🕗 Version & Regression Information
5.2.0-dev.20230621

### 💻 Repro

```ts
const a = {}; // run inline variable on `a`

a.toString();
```

### Bug
This results in invalid code:

```ts
{}.toString();
```

### PR DESCRIPTION
**Pull request by MariaSolOs**
Fixes #54736

### CODE REVIEW

**Review by Josh-Cena — COMMENTED** [review]
For object literals, the parentheses are only required when it appears at the beginning of an expression statement or concise arrow body; for number literals, the parentheses are only required when it's used in a dotted access and the number is base-10 without a decimal point. In addition, I don't think other literals should ever be parenthesized.

More suggested test cases:

```js
const f = function () {};
f.call();

const f = async function () {};
f.call();

const f = class {};
f.call();

// You may decide this is not worth fixing
const l = let[0];
l.call();
```

**Comment by MariaSolOs** [comment]
> For object literals, the parentheses are only required when it appears at the beginning of an expression statement or concise arrow body; for number literals, the parentheses are only required when it's used in a dotted access and the number is base-10 without a decimal point. In addition, I don't think other literals should ever be parenthesized.
> 
> More suggested test cases:
> 
> ```js
> const f = function () {};
> f.call();
> 
> const f = async function () {};
> f.call();
> 
> const f = class {};
> f.call();
> 
> // You may decide this is not worth fixing
> const l = let[0];
> l.call();
> ```

Yeah, I didn’t mean to publish the PR yet :)

**Comment by DanielRosenwasser** [comment]
@rbuckton I have a hard time believing we don't have some shared logic for these sorts of transformations. Any clues if there's something shared in services to do this?

**Comment by rbuckton** [comment]
> @rbuckton I have a hard time believing we don't have some shared logic for these sorts of transformations. Any clues if there's something shared in services to do this?

In our node factory we apply automatic parenthesization logic for these cases, so we don't have to worry about it. However, this is using our `textChanges` API in services which doesn't use the auto-parenthesization logic when you are inserting a node into what is essentially an arbitrary position in the source text.

It would be possible to leverage our auto-parenthesization logic in `textChanges`, though, since we can call into the parenthesizer rules outside of transforms (such as in the printer), and `textChanges`/`ChangeTracker` could probably determine the relevant parent node for the change and use that to pick the correct rule.

For an example of how this is used in the printer, look at `maybeEmitExpression` in emitter.ts.

**Comment by MariaSolOs** [comment]
It sounds like reusing the parenthesizing logic would require a larger refactoring that could be part of a separate PR. Thoughts @DanielRosenwasser @rbuckton?

File: tests/cases/fourslash/inlineVariableParensPropertyAccess.ts [hunk 1]
```diff
@@ -1,27 +1,27 @@
...
-    newContent: `const y = (1 + 2) * 3;
-(() => { })();`
-});
\ No newline at end of file
+/// <reference path="fourslash.ts" />
+
+////const /*a1*/foo/*b1*/ = {};
+////foo.toString();
+////const /*a2*/bar/*b2*/ = 0;
+////bar.toFixed().toString();
```
Comment (DanielRosenwasser) [inline tests/cases/fourslash/inlineVariableParensPropertyAccess.ts:6]: Can you add a test case for when the number already has a `.` in it:

```suggestion
////const /*a3*/baz/*b3*/ = 3.141592;
////baz.toPrecision(3);
```

Also, a test for other literals like strings and booleans?

File: src/services/refactors/inlineVariable.ts [hunk 2]
```diff
@@ -228,7 +231,13 @@ function getReplacementExpression(reference: Node, replacement: Expression): Exp
...
     // Functions also need to be parenthesized.
     // E.g.: const f = () => {}; f(); -> (() => {})();
-    if (isFunctionLike(replacement) && isCallLikeExpression(parent)) {
+    if (isFunctionLike(replacement) && (isCallLikeExpression(parent) || isPropertyAccessExpression(parent))) {
+        return factory.createParenthesizedExpression(replacement);
+    }
+
+    // Property access of literals need parentheses.
+    // E.g.: const x = 1; x.toString(); -> (1).toString();
+    if (isPropertyAccessExpression(parent) && (isLiteralTypeLikeExpression(replacement) || isObjectLiteralExpression(replacement))) {
```
Comment (DanielRosenwasser) [inline src/services/refactors/inlineVariable.ts:240]: `isLiteralTypeLikeExpression` is not the right helper. Literal type-likes correspond to nodes that can be used in type position that are literals. You are more concerned with nodes in the expression space, and it just so happens that we use the same kinds of nodes in both places.

As far as I remember for property accesses, decimal numeric literals with no exponent or fractional component are really the only literals where you need to parenthesize.

```ts
"hello".toString(); // ✅
true.toString();    // ✅
false.toString();   // ✅
1.0.toString();     // ✅
1e32.toString();    // ✅
0xcafe.toString();  // ✅
```

But you can just say "only parenthesize when the node is a numeric literal" just to make this simpler.

**Review by DanielRosenwasser — CHANGES_REQUESTED** [review]
I'm okay with doing a simpler check for now. I left some feedback though first.

**Comment by rbuckton** [comment]
> It sounds like reusing the parenthesizing logic would require a larger refactoring that could be part of a separate PR. Thoughts @DanielRosenwasser @rbuckton?

I think the parenthesization logic you need is basically this, though it needs testing:

```ts
/** @internal */
export function parenthesizeReplacementExpression(parenthesizer: ParenthesizerRules, parent: Node, target: Expression, replacement: Expression) {
    switch (parent.kind) {
        case SyntaxKind.BinaryExpression:
            return target === (parent as BinaryExpression).left ?
                parenthesizer.parenthesizeLeftSideOfBinary((parent as BinaryExpression).operatorToken.kind, replacement) :
                parenthesizer.parenthesizeRightSideOfBinary((parent as BinaryExpression).operatorToken.kind, (parent as BinaryExpression).left, replacement);
        case SyntaxKind.ComputedPropertyName:
            return parenthesizer.parenthesizeExpressionOfComputedPropertyName(replacement);
        case SyntaxKind.ConditionalExpression:
            if (target === (parent as ConditionalExpression).condition) return parenthesizer.parenthesizeConditionOfConditionalExpression(replacement);
            return parenthesizer.parenthesizeBranchOfConditionalExpression(replacement);
        case SyntaxKind.ExportAssignment:
            return (parent as ExportAssignment).isExportEquals ? replacement :
                parenthesizer.parenthesizeExpressionOfExportDefault(replacement);
        case SyntaxKind.NewExpression:
            return (parent as NewExpression).expression === target ?
                parenthesizer.parenthesizeExpressionOfNew(replacement) :
                parenthesizer.parenthesizeExpressionForDisallowedComma(replacement);
        case SyntaxKind.Decorator:
        case SyntaxKind.PropertyAccessExpression:
            return parenthesizer.parenthesizeLeftSideOfAccess(replacement);
        case SyntaxKind.ElementAccessExpression:
            return (parent as ElementAccessExpression).expression === target ?
                parenthesizer.parenthesizeLeftSideOfAccess(replacement) :
                parenthesizer.parenthesizeExpressionForDisallowedComma(replacement);
        case SyntaxKind.CallExpression:
            return (parent as CallExpression).expression === target ?
                parenthesizer.parenthesizeLeftSideOfAccess(replacement) :
                parenthesizer.parenthesizeExpressionForDisallowedComma(replacement);
        case SyntaxKind.TaggedTemplateExpression:
        case SyntaxKind.ExpressionWithTypeArguments:
        case SyntaxKind.NonNullExpression:
            return parenthesizer.parenthesizeLeftSideOfAccess(replacement);
        case SyntaxKind.PostfixUnaryExpression:
            return parenthesizer.parenthesizeOperandOfPostfixUnary(replacement);
        case SyntaxKind.TypeAssertionExpression:
        case SyntaxKind.DeleteExpression:
        case SyntaxKind.TypeOfExpression:
        case SyntaxKind.VoidExpression:
        case SyntaxKind.AwaitExpression:
        case SyntaxKind.PrefixUnaryExpression:
            return parenthesizer.parenthesizeOperandOfPrefixUnary(replacement);
        case SyntaxKind.ArrayLiteralExpression:
        case SyntaxKind.YieldExpression:
        case SyntaxKind.SpreadElement:
        case SyntaxKind.SpreadAssignment:
            return parenthesizer.parenthesizeExpressionForDisallowedComma(replacement);
        case SyntaxKind.ForOfStatement:
        case SyntaxKind.ForInStatement:
        case SyntaxKind.SwitchStatement:
        case SyntaxKind.CaseClause:
            return (parent as ForInOrOfStatement | SwitchStatement | CaseClause).expression === target ?
                parenthesizer.parenthesizeExpressionForDisallowedComma(replacement) :
                replacement;
        case SyntaxKind.ShorthandPropertyAssignment:
            return (parent as ShorthandPropertyAssignment).objectAssignmentInitializer === target ?
                parenthesizer.parenthesizeExpressionForDisallowedComma(replacement) :
                replacement;
        case SyntaxKind.PropertyAssignment:
        case SyntaxKind.EnumMember:
        case SyntaxKind.Parameter:
        case SyntaxKind.PropertyDeclaration:
        case SyntaxKind.BindingElement:
        case SyntaxKind.VariableDeclaration:
            return (parent as ParameterDeclaration | PropertyDeclaration | PropertyAssignment | BindingElement | VariableDeclaration | EnumMember).initializer === target ?
                parenthesizer.parenthesizeExpressionForDisallowedComma(replacement) :
                replacement;
        case SyntaxKind.ExpressionStatement:
            return parenthesizer.parenthesizeExpressionOfExpressionStatement(replacement);
        case SyntaxKind.ArrowFunction:
            return parenthesizer.parenthesizeConciseBodyOfArrowFunction(replacement);
    }
    return replacement;
}
```

It's up to you if you want to try this here or take it as a separate PR.

**Comment by MariaSolOs** [comment]
@rbuckton in the snippet above, is `parent` the `parent` of `target`?

**Comment by rbuckton** [comment]
> @rbuckton in the snippet above, is `parent` the `parent` of `target`?

Yes.

**Comment by MariaSolOs** [comment]
@rbuckton I'm testing `parenthesizeReplacementExpression` and I think (?) there's a bug in `parenthesizeLeftSideOfAccess`. If the expression is a function expression, then it is not parenthesized, despite something like `function () {}.call(null)` being invalid. Is this a bug or intentional?

**Comment by rbuckton** [comment]
Not a bug, but a caveat. FunctionExpression is a PrimaryExpression, so `function () {}.call()` *is* legal, *except* when it's the expression of an ExpressionStatement. This would normally be caught by the NodeFactory as the AST is rebuilt during a transform, but since you aren't doing a full transform, you will still have to watch out for that one.

**Comment by rbuckton** [comment]
> Not a bug, but a caveat. FunctionExpression is a PrimaryExpression, so `function () {}.call()` _is_ legal, _except_ when it's the expression of an ExpressionStatement. This would normally be caught by the NodeFactory as the AST is rebuilt during a transform, but since you aren't doing a full transform, you will still have to watch out for that one.

You will also have to verify the parent is not `export default`, since `export default function` can only be a FunctionDeclaration, so:

```ts
// before inline 'value'
const value = function () { }.call();
export default value;

// after inline 'value'
export default (function () { }.call());
```

Also, the same two restrictions for `function` apply to `class` as well, so a ClassExpression must also be parenthesized if inlined into an ExpressionStatement or `export default`:

```js
// before inline 'value'
const value = class { static f() {} }.f();
value;
// or
export default value;

// after inline 'value'
(class { static f() {} }.f());
// or
export default (class { static f() {} }.f());
```

I'll look through the parenthesizer rules to see if there are any other cases where the grandparent (or higher) might indicate parenthesization is required.

**Comment by MariaSolOs** [comment]
> You will also have to verify the parent is not export default

I don't allow exported variables to be inlined, and so I don't think we should worry about this.

**Comment by rbuckton** [comment]
> > You will also have to verify the parent is not export default
> 
> I don't allow exported variables to be inlined, and so I don't think we should worry about this.

Even `export default value`? It's not the binding `value` that is exported, so I'm not sure it is caught by `isDeclarationExported`.

**Comment by rbuckton** [comment]
Here's another case to verify:
```js
// before inline `a`
const a = x.y();
new a.b();

// after inline `a`, `x.y()` needs parens:
new (x.y()).b();
```

Here, we parent of `a` (`a.b`) isn't sufficient to verify parenthesis since `new x.y().b()` has a different meaning than `new (x.y()).b()`.

Another way to check if there are other necessary parenthesis to add would be to do something like this:

```ts
let currentParent: Node | undefined = parent;
let currentTarget: Node = target;
let currentReplacement: Node = replacement;
while (currentParent) {
    // update `currentParent`, replacing `currentTarget` with `currentReplacement`
    const updatedParent = visitEachChild(currentParent, child => child === currentTarget ? currentReplacement : child, nullTransformationContext);

    // visit `updatedParent` and look to see if we find a parenthesized child that contains `currentReplacement`
    let needsParens = false;
    visitEachChild(updatedParent, child => {
        // NOTE: unfortunately, we can't break early, but we can skip the check if we found a match.
        // NOTE: in the future, we could also check for `isParenthesizedType` if we decide to support inlining type variables.
        if (!needsParens && isParenthesizedExpression(child) && child.expression === currentReplacement) {
            needsParens = true;
        }
        return child;
    }, nullTransformationContext);

    if (needsParens) {
        // TODO: inject parens around `currentReplacement` in output
    }

    // Walk up the spine to check for other parenthesis inconsistencies
    currentTarget = currentParent;
    currentReplacement = updatedParent;
    currentParent = currentTarget.parent;
}
```

**Comment by MariaSolOs** [comment]
@rbuckton thank you for the hints! However, now that I think about it, I'm inclined to leave the code as it is for now. The main motivation to use the parenthesizer rules was to lift common logic/code, but it seems like there are still several cases I need to handle separately. There are also several cases which I don't need to worry about in this refactor since I restrict its availability.

What do you think (cc @DanielRosenwasser)?

**Review by DanielRosenwasser — APPROVED** [review]
Let's roll with something simple for now.

**Comment by DanielRosenwasser** [comment]
FWIW you may want to add Ron's test cases if you feel like it could add better coverage.

**Pull request merged**